Base/QTGUI/qSlicerApplication.cxx

```diff
@@ -396,10 +397,29 @@ bool qSlicerApplication::notify(QObject *receiver, QEvent *event)
   catch( std::exception& exception )
     {
     QString errorMessage;
-    errorMessage = tr("%1 has caught an internal error.\n\n").arg(this->applicationName());
-    errorMessage += tr("You may be able to continue from this point, but results are undefined.\n\n");
-    errorMessage += tr("Suggested action is to save your work and restart.");
-    errorMessage += tr("\n\nIf you have a repeatable sequence of steps that causes this message, ");
+    errorMessage = tr("%1 has caught an application error, ").arg(this->applicationName());
+    errorMessage += tr("please save your work and restart.\n\n");
+    if (!tr(exception.what()).compare(std::bad_alloc().what()))
```

Comment:
Comparing the strings would probably work, but it's not really the right thing to do.  Instead we should provide a specific catch body for std::bad_alloc and then a generic catch(...) body.

I didn't know exactly how this worked either, but I looked it up and wrote this for testing.  If you change around the order of the statements in the body of the try block you can see how it catches the various exceptions.

```c++
#include <exception>
#include <iostream>

int main()
{
  char *buffer = nullptr;
  try {
    unsigned long billion = 1000 * 1000 * 1000;
    buffer = new char[100000 * billion ];
    std::exception std_exception;
    throw 0;
    throw std_exception;
  } catch (std::bad_alloc e) {
    std::cerr << "caught bad alloc" << std::endl;
    std::cerr << e.what() << std::endl;
    exit(1);
  } catch (std::exception e) {
    std::cerr << "generic exception" << std::endl;
    std::cerr << e.what() << std::endl;
    exit(2);
  } catch (...) {
    std::cerr << "something odd happened" << std::endl;
    exit(2);
  }
  delete [] buffer;
  std::cout << "no problems" << std::endl;
  exit(0);
}
```

Reply:
Hi Steve, thanks for your comment. I've added the specific catch for std::bad_alloc before the generic catch as suggested. Some repeated codes are generated due to the addition, but I think the current form maintains good readability. I didn't add the last "catch ( ... )" as in the example since other modules may deliberately throw event for their own purpose.  Let me know what you think. Thanks!